File: settlement/room/law/prison/PrisonInstance.java
package settlement.room.law.prison;

import game.GAME;
import init.constant.C;
import init.resources.RBIT;
import init.resources.RBIT.RBITImp;
import init.resources.RESOURCES;
import init.resources.ResGEat;
import init.type.CAUSE_LEAVES;
import settlement.entity.ENTITY;
import settlement.entity.humanoid.HEvent;
import settlement.entity.humanoid.Humanoid;
import settlement.entity.humanoid.ai.types.prisoner.AIModule_Prisoner;
import settlement.main.SETT;
import settlement.misc.job.JOBMANAGER_HASER;
import settlement.misc.job.JOB_MANAGER;
import settlement.misc.job.SETT_JOB;
import settlement.room.industry.module.Industry;
import settlement.room.industry.module.ROOM_PRODUCER_INSTANCE;
import settlement.room.main.RoomInstance;
import settlement.room.main.TmpArea;
import settlement.room.main.furnisher.FurnisherItem;
import settlement.room.main.job.JobPositions;
import settlement.room.main.util.RoomInit;
import settlement.stats.STATS;
import snake2d.Renderer;
import snake2d.util.bit.Bits;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.Coo;
import snake2d.util.datatypes.DIR;
import snake2d.util.misc.CLAMP;
import util.rendering.RenderData;
import util.rendering.ShadowBatch;

final class PrisonInstance extends RoomInstance implements JOBMANAGER_HASER, ROOM_PRODUCER_INSTANCE {

	private static final long serialVersionUID = 1L;
	final Jobs jobs;

	
	private short prisoners = 0;
	
	private static final Bits bprisoners = new Bits(0b01111);
	private final short[] cellsXY;
	private short cellI = 0;
	final RBITImp fetch = new RBITImp();
	private long[] productionData;
	float riotChance = 1;
	boolean hasWarned = false;
	
	protected PrisonInstance(ROOM_PRISON b, TmpArea area, RoomInit init) {
		super(b, area, init);
		
		int cells = 0;
		for (COORDINATE c : body()) {
			if (is(c)) {
				candle(c.x(), c.y());
				if (SETT.ROOMS().fData.tileData.get(c) == Constructor.CODE_ENTRANCE) {
					cells++;
				}
			}
		}
		
		cellsXY = new short[cells*2];
		cells = 0;
		for (COORDINATE c : body()) {
			if (is(c)) {
				if (SETT.ROOMS().fData.tileData.get(c) == Constructor.CODE_ENTRANCE) {
					cellsXY[cells++] = (short) c.x();
					cellsXY[cells++] = (short) c.y();
				}
			}
		}
		productionData = blueprintI().indu.makeData();
		jobs = new Jobs(this);
		
		int am = (int)Math.ceil(b.constructor.guards.get(this));
		employees().maxSet(am);
		employees().neededSet(am);
		
		for (ResGEat e : RESOURCES.EDI().all())
			if (e.serve)
				fetch.or(e.resource);
		activate();
		jobs.setAlwaysNew();
		
	}
	
	@Override
	protected void loadFix() {
		productionData = blueprintI().indu.makeDataFix(productionData);
		jobs.setAlwaysNew();
		jobs.resNotFound.clear();
	}
	
	void candle(int tx, int ty) {
		
		if (SETT.LIGHTS().is(tx, ty)) {
			SETT.LIGHTS().remove(tx, ty);
			FurnisherItem it = SETT.ROOMS().fData.item.get(tx, ty);
					
			for (DIR d : DIR.ORTHO) {
				if (SETT.ROOMS().fData.item.is(tx, ty, d, it)) {
					SETT.LIGHTS().candle(tx, ty, d.x()*(C.TILE_SIZEH-4), d.y()*(C.TILE_SIZEH-4));
					return;
				}
			}
		}
	}
	
	public int prisoners() {
		return prisoners;
	}
	
	public int prisonersMax() {
		return blueprintI().constructor.PRISONERS_PER_CELL*cellsXY.length/2;
	}
	
	@Override
	protected boolean render(Renderer r, ShadowBatch shadowBatch, RenderData.RenderIterator it) {
		it.lit();
		return super.render(r, shadowBatch, it);
	}

	@Override
	protected void activateAction() {
		blueprintI().incPrisoners(prisoners, prisonersMax());
	}

	@Override
	protected void deactivateAction() {
		for (ENTITY e : SETT.ENTITIES().getAllEnts()) {
			if (e != null && e instanceof Humanoid) {
				Humanoid h = (Humanoid) e;
				HEvent.Handler.removeRoom(h, this);
			}
		}
		
		for (int i = 0; i < cellsXY.length; i+=2) {
			cellI += 2;
			if (cellI >= cellsXY.length)
				cellI = 0;
			int tx = cellsXY[cellI];
			int ty = cellsXY[cellI+1];
			int data = SETT.ROOMS().data.get(tx, ty);
			data = bprisoners.set(data, 0);
			SETT.ROOMS().data.set(this, tx, ty, data);
		}
		
		blueprintI().incPrisoners(-prisoners, -prisonersMax());
		prisoners = 0;
	}

	@Override
	protected void updateAction(double updateInterval, boolean day) {
		blueprintI().indu.updateRoom(this);
		jobs.searchAgain();
		if (day && prisoners() > 0) {
			
			float prev = riotChance;
			
			double v = (double)employees().employed()/employees().max() - 1;
			
			
			RBIT.RBITImp.tmp.clearSet(fetch);
			RBIT.RBITImp.tmp.xor(jobs.resNotFound);
			if (RBIT.RBITImp.tmp.isClear())
				v -= 1;
			
			
			v /= 4;
			
			if (v == 0) {
				riotChance = 1;
				return;
			}
				
			
			riotChance += v;
			riotChance = (float) CLAMP.d(riotChance, 0, 1);
			if (riotChance < 0.5 && riotChance < prev && !hasWarned) {
				Gui.mWarn(this);
				hasWarned = true;
			}else if (riotChance <= 0) {
				Gui.m(this);
				hasWarned = false;
				riotChance = 1;
				for (ENTITY e : SETT.ENTITIES().getAllEnts()) {
					if (e instanceof Humanoid) {
						Humanoid a = (Humanoid) e;
						if (AIModule_Prisoner.isPrisoner(a, this)) {
							STATS.LAW().escapeInc();
							a.kill(false, CAUSE_LEAVES.OTHER());
						}		
					}
				}
				
				
			}
		}
		

	}
	
	@Override
	public JOB_MANAGER getWork() {
		return jobs;
	}
	
	@Override
	protected void dispose() {
		
		
	}

	@Override
	public ROOM_PRISON blueprintI() {
		return (ROOM_PRISON) blueprint();
	}
	
	COORDINATE registerPrisoner(COORDINATE c) {
		if (prisoners >= prisonersMax())
			throw new RuntimeException();
		if (!active())
			throw new RuntimeException();
		
		if (is(c) && SETT.ROOMS().fData.tileData.get(c) == Constructor.CODE_ENTRANCE) {

			int tx = c.x();
			int ty = c.y();
			int data = SETT.ROOMS().data.get(tx, ty);
			if (bprisoners.get(data) < blueprintI().constructor.PRISONERS_PER_CELL) {
				incPrisoner(tx, ty, 1);
				Coo.TMP.set(tx, ty);
				return Coo.TMP;
			}
		}
		
		int pris = 0;
		
		for (int i = 0; i < cellsXY.length; i+=2) {
			cellI += 2;
			if (cellI >= cellsXY.length)
				cellI = 0;
			int tx = cellsXY[cellI];
			int ty = cellsXY[cellI+1];
			int data = SETT.ROOMS().data.get(tx, ty);
			pris += bprisoners.get(data);
			if (bprisoners.get(data) < blueprintI().constructor.PRISONERS_PER_CELL) {
				incPrisoner(tx, ty, 1);
				Coo.TMP.set(tx, ty);
				return Coo.TMP;
			}
		}
		
		throw new RuntimeException(pris + " " + prisoners + " " + prisonersMax());
	}
	
	void incPrisoner(int tx, int ty, int am) {
		int data = SETT.ROOMS().data.get(tx, ty);
		if (am < 0) {
			if (bprisoners.get(data) + am < 0 || prisoners < 0) {
				return;
			}
		}else if (am > 0) {
			if (bprisoners.get(data) + am > blueprintI().constructor.PRISONERS_PER_CELL || prisoners > prisonersMax()) {
				GAME.Error("prison " + (prisoners > prisonersMax()));
				return;
			}
		}
		prisoners += am;
		data = bprisoners.inc(data, am);
		SETT.ROOMS().data.set(this, tx, ty, data);
		blueprintI().incPrisoners(am, 0);
	}
	
	void removePrisoner(int tx, int ty) {
		if (!is(tx, ty))
			return;
		if (SETT.ROOMS().fData.tileData.get(tx, ty) != Constructor.CODE_ENTRANCE)
			return;
		int data = SETT.ROOMS().data.get(tx, ty);
		if (bprisoners.get(data) == 0) {
			return;
		}
		incPrisoner(tx, ty, -1);
	}
	
	boolean isReserved(int tx, int ty) {
		if (SETT.ROOMS().fData.tileData.get(tx, ty) != Constructor.CODE_ENTRANCE)
			return false;
		int data = SETT.ROOMS().data.get(tx, ty);
		if (bprisoners.get(data) == 0) {
			return false;
		}
		return true;
	}
	
	static class Jobs extends JobPositions<PrisonInstance> {

		private static final long serialVersionUID = 1L;

		public Jobs(PrisonInstance ins) {
			super(ins);
			setAlwaysNew();
		}

		@Override
		protected SETT_JOB get(int tx, int ty) {
			if (Food.init(tx, ty) != null) {
				return Food.init(tx, ty);
			}
			if (Latrine.init(tx, ty) != null)
				return Latrine.init(tx, ty);
			if (Cell.init(tx, ty) != null)
				return Cell.init(tx, ty);
			return null;
		}

		@Override
		protected boolean isAndInit(int tx, int ty) {
			if (Food.init(tx, ty) != null)
				return true;
			if (Latrine.init(tx, ty) != null)
				return true;
			if (Cell.init(tx, ty) != null)
				return true;
			return false;
			
			
		}
	}

	@Override
	public long[] productionData() {
		return productionData;
	}

	@Override
	public Industry industry() {
		return blueprintI().indu;
	}

	@Override
	public int industryI() {
		return 0;
	}

}

File: LegacySource/settlement/room/law/prison/PrisonInstance.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Settlement.Room.Law.Prison
{
    public class PrisonInstance : RoomInstance, JOBMANAGER_HASER, ROOM_PRODUCER_INSTANCE
    {
        private static readonly long serialVersionUID = 1L;
        public Jobs jobs;

        private short prisoners = 0;

        private static readonly Bits bprisoners = new Bits(0b01111);
        private readonly short[] cellsXY;
        private short cellI = 0;
        private readonly RBITImp fetch = new RBITImp();
        private long[] productionData;
        private float riotChance = 1;
        private bool hasWarned = false;

        protected PrisonInstance(ROOM_PRISON b, TmpArea area, RoomInit init) : base(b, area, init)
        {
            int cells = 0;
            foreach (var c in body())
            {
                if (Is(c))
                {
                    candle(c.x, c.y);
                    if (SETT.ROOMS().fData.tileData.Get(c) == Constructor.CODE_ENTRANCE)
                    {
                        cells++;
                    }
                }
            }

            cellsXY = new short[cells * 2];
            cells = 0;
            foreach (var c in body())
            {
                if (Is(c))
                {
                    if (SETT.ROOMS().fData.tileData.Get(c) == Constructor.CODE_ENTRANCE)
                    {
                        cellsXY[cells++] = (short)c.x;
                        cellsXY[cells++] = (short)c.y;
                    }
                }
            }
            productionData = blueprintI().indu.MakeData();
            jobs = new Jobs(this);

            int am = (int)Math.Ceiling(b.constructor.guards.Get(this));
            employees().maxSet(am);
            employees().neededSet(am);

            foreach (var e in RESOURCES.EDI().All())
                if (e.serve)
                    fetch.Or(e.resource);
            Activate();
            jobs.SetAlwaysNew();
        }

        protected override void LoadFix()
        {
            productionData = blueprintI().indu.MakeDataFix(productionData);
            jobs.SetAlwaysNew();
            jobs.resNotFound.Clear();
        }

        void candle(int tx, int ty)
        {
            if (SETT.LIGHTS().Is(tx, ty))
            {
                SETT.LIGHTS().Remove(tx, ty);
                FurnisherItem it = SETT.ROOMS().fData.item.Get(tx, ty);

                foreach (DIR d in DIR.ORTHO)
                {
                    if (SETT.ROOMS().fData.item.Is(tx, ty, d, it))
                    {
                        SETT.LIGHTS().Candle(tx, ty, d.x * (C.TILE_SIZEH - 4), d.y * (C.TILE_SIZEH - 4));
                        return;
                    }
                }
            }
        }

        public int Prisoners()
        {
            return prisoners;
        }

        public int PrisonersMax()
        {
            return blueprintI().constructor.PRISONERS_PER_CELL * cellsXY.Length / 2;
        }

        protected override bool Render(Renderer r, ShadowBatch shadowBatch, RenderData.RenderIterator it)
        {
            it.lit();
            return base.Render(r, shadowBatch, it);
        }

        protected override void ActivateAction()
        {
            blueprintI().IncPrisoners(prisoners, prisonersMax());
        }

        protected override void DeactivateAction()
        {
            foreach (var e in SETT.ENTITIES().GetAllEnts())
            {
                if (e != null && e is Humanoid)
                {
                    Humanoid h = (Humanoid)e;
                    HEvent.Handler.RemoveRoom(h, this);
                }
            }

            for (int i = 0; i < cellsXY.Length; i += 2)
            {
                cellI += 2;
                if (cellI >= cellsXY.Length)
                    cellI = 0;
                int tx = cellsXY[cellI];
                int ty = cellsXY[cellI + 1];
                int data = SETT.ROOMS().data.Get(tx, ty);
                data = bprisoners.Set(data, 0);
                SETT.ROOMS().data.Set(this, tx, ty, data);
            }

            blueprintI().IncPrisoners(-prisoners, -prisonersMax());
            prisoners = 0;
        }

        protected override void UpdateAction(double updateInterval, bool day)
        {
            blueprintI().indu.UpdateRoom(this);
            jobs.SearchAgain();
            if (day && prisoners() > 0)
            {
                float prev = riotChance;

                double v = (double)employees().Employed() / employees().Max() - 1;

                RBIT.RBITImp.tmp.ClearSet(fetch);
                RBIT.RBITImp.tmp.Xor(jobs.resNotFound);
                if (RBIT.RBITImp.tmp.IsClear())
                    v -= 1;

                v /= 4;

                if (v == 0)
                {
                    riotChance = 1;
                    return;
                }

                riotChance += (float)v;
                riotChance = (float)CLAMP.d(riotChance, 0, 1);
                if (riotChance < 0.5 && riotChance < prev && !hasWarned)
                {
                    Gui.mWarn(this);
                    hasWarned = true;
                }
                else if (riotChance <= 0)
                {
                    Gui.m(this);
                    hasWarned = false;
                    riotChance = 1;
                    foreach (var e in SETT.ENTITIES().GetAllEnts())
                    {
                        if (e is Humanoid)
                        {
                            Humanoid a = (Humanoid)e;
                            if (AIModule_Prisoner.IsPrisoner(a, this))
                            {
                                STATS.LAW().escapeInc();
                                a.kill(false, CAUSE_LEAVES.OTHER());
                            }
                        }
                    }
                }
            }
        }

        public JOB_MANAGER GetWork()
        {
            return jobs;
        }

        protected override void Dispose()
        {
        }

        public ROOM_PRISON BlueprintI()
        {
            return (ROOM_PRISON)blueprint();
        }

        COORDINATE RegisterPrisoner(COORDINATE c)
        {
            if (prisoners >= PrisonersMax())
                throw new RuntimeException();
            if (!active())
                throw new RuntimeException();

            if (Is(c) && SETT.ROOMS().fData.tileData.Get(c) == Constructor.CODE_ENTRANCE)
            {
                int tx = c.x;
                int ty = c.y;
                int data = SETT.ROOMS().data.Get(tx, ty);
                if (bprisoners.Get(data) < blueprintI().constructor.PRISONERS_PER_CELL)
                {
                    IncPrisoner(tx, ty, 1);
                    Coo.TMP.Set(tx, ty);
                    return Coo.TMP;
                }
            }

            int prisonersInCells = 0;
            for (int i = 0; i < cellsXY.Length; i += 2)
            {
                int tx = cellsXY[i];
                int ty = cellsXY[i + 1];
                int data = SETT.ROOMS().data.Get(tx, ty);
                prisonersInCells += bprisoners.Get(data);
            }

            if (prisonersInCells < PrisonersMax())
            {
                for (int i = 0; i < cellsXY.Length; i += 2)
                {
                    int tx = cellsXY[i];
                    int ty = cellsXY[i + 1];
                    int data = SETT.ROOMS().data.Get(tx, ty);
                    if (bprisoners.Get(data) < blueprintI().constructor.PRISONERS_PER_CELL)
                    {
                        IncPrisoner(tx, ty, 1);
                        Coo.TMP.Set(tx, ty);
                        return Coo.TMP;
                    }
                }
            }

            throw new RuntimeException("No available cells for prisoner");
        }

        void IncPrisoner(int tx, int ty, int am)
        {
            if (am < 0 && (bprisoners.Get(SETT.ROOMS().data.Get(tx, ty)) + am < 0 || prisoners < 0))
                return;
            if (am > 0 && (bprisoners.Get(SETT.ROOMS().data.Get(tx, ty)) + am > blueprintI().constructor.PRISONERS_PER_CELL || prisoners > PrisonersMax()))
            {
                GAME.Error("prison " + (prisoners > PrisonersMax()));
                return;
            }
            prisoners += am;
            int data = bprisoners.Inc(SETT.ROOMS().data.Get(tx, ty), am);
            SETT.ROOMS().data.Set(this, tx, ty, data);
            blueprintI().IncPrisoners(am, 0);
        }

        void RemovePrisoner(int tx, int ty)
        {
            if (!Is(tx, ty))
                return;
            if (SETT.ROOMS().fData.tileData.Get(tx, ty) != Constructor.CODE_ENTRANCE)
                return;
            int data = SETT.ROOMS().data.Get(tx, ty);
            if (bprisoners.Get(data) == 0)
                return;
            IncPrisoner(tx, ty, -1);
        }

        bool IsReserved(int tx, int ty)
        {
            if (SETT.ROOMS().fData.tileData.Get(tx, ty) != Constructor.CODE_ENTRANCE)
                return false;
            int data = SETT.ROOMS().data.Get(tx, ty);
            if (bprisoners.Get(data) == 0)
                return false;
            return true;
        }

        public class Jobs : JobPositions<PrisonInstance>
        {
            private static readonly long serialVersionUID = 1L;

            public Jobs(PrisonInstance ins) : base(ins)
            {
                SetAlwaysNew();
            }

            protected override SETT_JOB Get(int tx, int ty)
            {
                if (Food.init(tx, ty) != null)
                    return Food.init(tx, ty);
                if (Latrine.init(tx, ty) != null)
                    return Latrine.init(tx, ty);
                if (Cell.init(tx, ty) != null)
                    return Cell.init(tx, ty);
                return null;
            }

            protected override bool IsAndInit(int tx, int ty)
            {
                if (Food.init(tx, ty) != null)
                    return true;
                if (Latrine.init(tx, ty) != null)
                    return true;
                if (Cell.init(tx, ty) != null)
                    return true;
                return false;
            }
        }

        public long[] ProductionData()
        {
            return productionData;
        }

        public Industry Industry()
        {
            return blueprintI().indu;
        }

        public int IndustryI()
        {
            return 0;
        }
    }
}